### ISSUE TITLE
[Scaffolder] Bitbucket/BitbucketServer publish actions fail to authenticate with Git

### ISSUE THREAD

**Issue by namco1992** [backstage/backstage#12586]
## Expected Behavior

When using the action `publish:bitbucketServer` or `publish:bitbucket`, it should be able to authenticate with the credentials provided and commit the code successfully.

## Actual Behavior

When using the action `publish:bitbucketServer` or `publish:bitbucket`, it will create the repo successfully via HTTP API but fail to commit the code via git.

## Steps to Reproduce

1. Create a publish action in your template
```
    - id: publish
      name: Publish
      action: publish:bitbucket
      input:
        description: This is ${{ parameters.project_name }}
        repoUrl: ${{ parameters.repoUrl }}
        defaultBranch: ${{ parameters.defaultBranch }}
```
2. Run your template

## Context

The current implementation of the Bitbucket/BitbucketServer publish action seems to be influenced by [this doc](https://support.atlassian.com/bitbucket-cloud/docs/use-oauth-on-bitbucket-cloud/):
https://github.com/backstage/backstage/blob/7b33fa1a2ab8f2cf6d0eb3c34626f36d2cef66b9/plugins/scaffolder-backend/src/scaffolder/actions/builtin/publish/bitbucket.ts#L379-L382

The `x-token-auth` is a fixed string.

However, it doesn't work with a self-hosted Bitbucket server. According to the [docs here](https://confluence.atlassian.com/bitbucketserver/personal-access-tokens-939515499.html), the username is required on BitbucketServer like:

```bash
git clone https://username:MDM0MjM5NDc2MDxxxxxxxxxxxxxxxxxxxxx@bitbucketserver.com/scm/projectname/teamsinspace.git
```

or this if you don't have the username:

```bash
git clone -c http.extraHeader='Authorization: Bearer MDM0MjM5NDc2MDxxxxxxxxxxxxxxxxxxxxx' https://bitbucketserver.com/scm/projectname/teamsinspace.git
```

I tested the above methods and it works with our Bitbucket server version 7.21.

If the current implementation was for a historical version of Bitbucket server, then we might need help for testing it out in other versions. If the current implementation is just taken from the Bitbucket cloud, then it should be fixed.

## Your Environment

- Bitbucket server version 7.21
- Output of `yarn backstage-cli info`: 

```text
OS:   Darwin 21.5.0 - darwin/arm64
node: v16.15.0
yarn: 1.22.18
cli:  0.17.2 (installed)

Dependencies:
  @backstage/app-defaults                            1.0.3
  @backstage/backend-common                          0.14.0
  @backstage/backend-tasks                           0.3.2
  @backstage/backend-test-utils                      0.1.25
  @backstage/catalog-client                          1.0.3
  @backstage/catalog-model                           1.0.3
  @backstage/cli-common                              0.1.9
  @backstage/cli                                     0.17.2
  @backstage/config-loader                           1.1.2
  @backstage/config                                  1.0.1
  @backstage/core-app-api                            1.0.3
  @backstage/core-components                         0.9.5
  @backstage/core-plugin-api                         1.0.3
  @backstage/dev-utils                               1.0.3
  @backstage/errors                                  1.0.0
  @backstage/integration-react                       1.1.1
  @backstage/integration                             1.2.1
  @backstage/plugin-api-docs                         0.8.6
  @backstage/plugin-app-backend                      0.3.33
  @backstage/plugin-auth-backend                     0.14.1
  @backstage/plugin-auth-node                        0.2.2
  @backstage/plugin-bitbucket-cloud-common           0.1.0
  @backstage/plugin-catalog-backend-module-bitbucket 0.2.0
  @backstage/plugin-catalog-backend                  1.2.0
  @backstage/plugin-catalog-common                   1.0.3
  @backstage/plugin-catalog-graph                    0.2.18
  @backstage/plugin-catalog-import                   0.8.9
  @backstage/plugin-catalog-react                    1.1.1
  @backstage/plugin-catalog                          1.3.0
  @backstage/plugin-kubernetes-common                0.3.0
  @backstage/plugin-kubernetes                       0.6.6
  @backstage/plugin-org                              0.5.6
  @backstage/plugin-pagerduty                        0.4.0
  @backstage/plugin-permission-common                0.6.2
  @backstage/plugin-permission-node                  0.6.2
  @backstage/plugin-permission-react                 0.4.2
  @backstage/plugin-proxy-backend                    0.2.27
  @backstage/plugin-scaffolder-common                1.1.1
  @backstage/plugin-search-backend-node              0.6.2
  @backstage/plugin-search-backend                   0.5.3
  @backstage/plugin-search-common                    0.3.5
  @backstage/plugin-search-react                     0.2.1
  @backstage/plugin-search                           0.9.0
  @backstage/plugin-techdocs-backend                 1.1.2
  @backstage/plugin-techdocs-module-addons-contrib   1.0.1
  @backstage/plugin-techdocs-node                    1.1.2
  @backstage/plugin-techdocs-react                   1.0.1
  @backstage/plugin-techdocs                         1.2.0
  @backstage/plugin-user-settings                    0.4.5
  @backstage/release-manifests                       0.0.4
  @backstage/test-utils                              1.1.1
  @backstage/theme                                   0.2.15
  @backstage/types                                   1.0.0
  @backstage/version-bridge                          1.0.1
✨  Done in 1.09s.
```

**Comment by benjdlambert** [comment]
Hmm - so the block of code you linked is only executed when passing in a token to the action, is that something that you're doing? 

I wonder if you try and set `username` and `token` in the `app-config.yaml` does that work?

**Comment by benjdlambert** [comment]
@pjungermann here for visibility too.

**Comment by pjungermann** [comment]
At some point I was splitting the action `publish:bitbucket` into `publish:bitbucketCloud` and `publish:bitbucketServer` as the integration was split. Theoretically it was usable for Bitbucket Server, but maybe it was never really support 🤔 

The authentication is done using OAuth. This is powered by the auth module/plugin for "bitbucket" which is Bitbucket Cloud only.

So.. maybe it is rather a feature request than a bug. But a bit unclear.

@dhruvaldarji @renlord Did you ever use the scaffolder with your Bitbucket Server setup?

**Comment by namco1992** [comment]
Hi @benjdlambert @pjungermann,

> Hmm - so the block of code you linked is only executed when passing in a token to the action, is that something that you're doing?

That's not a good example because it actually works if you config `integrationConfig.config.username`, I think it's clearer here:

https://github.com/backstage/backstage/blob/7b33fa1a2ab8f2cf6d0eb3c34626f36d2cef66b9/plugins/scaffolder-backend/src/scaffolder/actions/builtin/publish/bitbucketServer.ts#L240-L243

It only takes a token and set the username to `x-token-auth`, and based on our tests, this way doesn't work for Bitbucket server.

> The authentication is done using OAuth. This is powered by the auth module/plugin for "bitbucket" which is Bitbucket Cloud only.

I think we might be talking about different parts, the auth here is used in the `initRepoAndPush` func to authenticate with git. I'm not sure what's the "auth module/plugin" you are referring to, if you mean this https://backstage.io/docs/auth/bitbucket/provider then I think it's not used in these actions.

> So.. maybe it is rather a feature request than a bug. But a bit unclear.

To summarize the issue here:
1. `publish:bitbucket` **will work if you set both `username` and `token`**, because the logic there tries to cater to both Bitbucket server and Bitbucket cloud, which makes the `username` and `token` combination possible.
https://github.com/backstage/backstage/blob/7b33fa1a2ab8f2cf6d0eb3c34626f36d2cef66b9/plugins/scaffolder-backend/src/scaffolder/actions/builtin/publish/bitbucket.ts#L378-L392

2. `publish:bitbucketServer` **won't work** (at least for Bitbucket server version 7.21) because it's fixed to `x-token-auth` as username.

**Comment by dhruvaldarji** [comment]
> @dhruvaldarji @renlord Did you ever use the scaffolder with your Bitbucket Server setup?

Unfortunately we haven't been able to implement the Scaffolder yet so I haven't been able to test any of these features.

**Comment by pjungermann** [comment]
@namco1992 as of the docs at `BitbucketIntegrationConfig.username` it was meant to be used by Bitbucket Cloud only. For Bitbucket Server, https://confluence.atlassian.com/bitbucketserver/personal-access-tokens-939515499.html was referenced for auth at `BitbucketIntegrationConfig.token`.

This was the reason for `BitbucketServerIntegrationConfig` to only contain `token`.

Based on what you describe, it seems like `username` + `password` is also used. Based on https://developer.atlassian.com/server/bitbucket/how-tos/example-basic-authentication/ seems like Basic Auth using these is an option.

https://developer.atlassian.com/server/bitbucket/how-tos/command-line-rest/#authentication mentions
> The REST APIs support basic authentication, OAuth and session authentication.

**Comment by thi2pra1** [comment]
This PR will save us here in our company
thank you so much
